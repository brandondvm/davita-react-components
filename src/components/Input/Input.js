import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

const Input = (props) => {
  // Props
  const { label = '', placeholder = '', onBlur, innerLabel, selectButtons, errorMessage, name, id, register, accessibiltity, type = 'text', header, characterMax, defaultValue, hideMax } = props;
  // State Hooks
  const [activeButton, setActiveButton] = useState((selectButtons && selectButtons.activeButton) || 'left');  
  const [focusedLabel, setFocusedLabel] = useState(false);
  const [value, setValue] = useState(null);

  useEffect(() => {
    if (defaultValue && !value) {
      setValue(defaultValue);
    }
  }, [defaultValue, value])

  const handleButtonClick = (activeButton, callback) => {
    setActiveButton(activeButton);
    callback && callback();
  };  

  const handleFocus = (value, e) => {
    setFocusedLabel(value);
    if (!value) {
      onBlur && onBlur(e.target.value);
    }
  };

  const handleInputChange = e => {
    if (placeholder === 'HH:MM') {
      const inputValue = e.target.value;
      if (inputValue.length > value.length) {
        if (inputValue.length === 2) {
          e.target.value = `${inputValue}:`;
        }
        if (inputValue.length === 3 && !inputValue.includes(':')) {
          e.target.value = `${value}:${inputValue[inputValue.length - 1]}`;
        }
      }
    }
    setValue(e.target.value);
  }

  const focusedLabelColor = errorMessage ? '#CE3616' : '#0069B1';
  const unfocusedLabelColor = errorMessage ? '#CE3616' : '#0d1221';

  return (
    <div className="input-component">      
      { header && (
        <div className="input-component_header">
          <i className={`di ${ header.className }`} />
          <p className="input-component_header_text">{ header.text }</p>
        </div>
      )}
      <div className="input-component_input-area">        
        {label && (
          <label
            htmlFor={id}
            style={{ color: focusedLabel ? focusedLabelColor : unfocusedLabelColor }}>
            {label}
          </label>
        )}
        {!label && (
          <label
            htmlFor={id}
            className={`placeholder${focusedLabel || value ? ' focused' : ''}${!focusedLabel && value ? ' regular' : ''}`}
            style={{ color: focusedLabel && focusedLabelColor }}
          >
            {placeholder}
          </label>
        )}
        <input
          id={id}
          name={name}
          className={`input-component_input ${errorMessage ? 'error-present' : ''} ${innerLabel ? 'inner-label-present' : ''} ${selectButtons ? 'select-buttons-present' : ''}`}
          type={type}
          placeholder={label ? placeholder : ''}
          onFocus={() => handleFocus(true)}
          onBlur={e => handleFocus(false, e)}
          onChange={handleInputChange}
          maxLength={characterMax}
          ref={register}
          aria-label={accessibiltity?.ariaLabel}
          aria-invalid={!!errorMessage}
          inputMode={ type === 'number' || placeholder === "HH:MM" ? 'decimal' : 'text' }
        />
        {
          (innerLabel && !selectButtons) && (
            <div 
              className={`input-component_input-area_unit ${!label ? 'input-component_input-area_unit_no-label' : ''} ${focusedLabel || value ? 'active' : ''}`}>
              {innerLabel}
            </div>
          )
        }
        {
          selectButtons && (
            <div className="input-component_input-area_buttons">
              <button
                type="button"
                className={activeButton === 'left' ? 'active-button' : ''}
                onClick={() => handleButtonClick('left', selectButtons.leftButtonOnClick)}
              >
                {selectButtons.leftButtonLabel}
              </button>
              <button
                type="button"
                className={activeButton === 'right' ? 'active-button' : ''}
                onClick={() => handleButtonClick('right', selectButtons.rightButtonOnClick)}
              >
                {selectButtons.rightButtonLabel}
              </button>
            </div>
          )
        }

      </div>
      <div className="input-component_feedback">
        { characterMax && !hideMax && <span className="input-component_char-counter">{value?.length || 0}/{characterMax}</span>}
        { errorMessage && <span>{ errorMessage }</span> }
      </div>
    </div>
  );
};

Input.propTypes = {
  label: PropTypes.string,
  placeholder: PropTypes.string,
  innerLabel: PropTypes.string,
  errorMessage: PropTypes.string,
  characterMax: PropTypes.string,
  selectButtons: PropTypes.shape({
    leftButtonLabel: PropTypes.string,
    leftButtonOnClick: PropTypes.func,
    rightButtonLabel: PropTypes.string,
    rightButtonOnClick: PropTypes.func,
  }),
  id: PropTypes.string.isRequired,
};

export default Input;
