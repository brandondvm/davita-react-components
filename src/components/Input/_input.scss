.input-component {
  &_header {
    align-items: center;
    display: flex;
    flex-direction: row;
    &_text {
      color: $onyx;
      font-family: $bold-font;
      font-size: 16px;
      margin: 0;
    }
    i {
      font-size: 20px;
      margin-left: 5px;
      width: 28px;
    }
  }
  &_input-area {
    position: relative;
    margin-bottom: 8px;

    &_buttons {
      background: rgba($inactivebutton, 0.12);
      border-radius: 4px;
      height: 32px;
      padding: 2px;
      position: absolute;
      right: 8px;
      top: 27px;
      width: 108px;

      button {
        background: rgba($inactivebutton, 0);
        border-radius: 4px;
        font-size: 14px;
        height: 100%;
        width: 50%;
      }
      button.active-button {
        background-color: $white;
        font-family: $bold-font;
        &:active,
        &:focus {
          outline: none;
        }
      }
    }

    &_unit {
      color: $inactivebutton;
      position: absolute;
      right: 8px;
      top: 27px;
      height: 32px;
      line-height: 18px;
      font-size: 13px;
      padding: 6px 0;
      visibility: hidden;
      &_no-label {
        top: 8px;
      }
      &.active {
        visibility: visible;
      }
    }
  }

  label {
    background: $white;
    color: $onyx;
    position: relative;
    top: 9px;
    left: 10px;
    padding: 0 1px;

    &.placeholder {
      color: $mediumdarkgray;
      font-size: 14px;
      position: absolute;
      top: 13px;
      pointer-events: none;
      transition: transform 150ms ease-out;
    }

    &.regular {
      text-transform: capitalize;
      color: $onyx;
    }

    &.focused {
      text-transform: capitalize;
      transform: translateY(-23px);
    }
  }

  &_input {
    border-radius: 4px;
    border-width: 1px;
    border-color: $lightgray;
    color: $darkgray;
    font-size: 15px;
    height: 45px;
    padding: 12px 10px;
    width: 100%;

    &:focus {
      border: 2px solid $blue;
      color: $darkgray;
      outline: none;
    }
  }

  input[type='time'] {
    font-family: $regular-font!important;
    &::-moz-clear,
    &::-webkit-clear-button,
    &::-webkit-calendar-picker-indicator, 
    &::-webkit-datetime-edit-ampm-field {
      display: none;
    }
  }

  // Fix for input shadow in Safari
  input[type='time'],
  input[type='number'],
  input[type='text'] {
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
  }

  &_input.error-present {
    border: 2px solid $red;

    &:focus {
      border: 2px solid $red;
      color: $darkgray;
    }
  }

  &_input.inner-label-present {
    padding-right: 45px;
  }

  &_input.select-buttons-present {
    padding-right: 120px;
  }

  &_feedback {
    display: flex;
    color: $lightgray;
    font-size: 11px;
    line-height: 13px;
    padding: 0 10px;
    position: relative;
    width: 100%;
  }

  &_char-counter {
    position: absolute;
    right: 10px;
  }
}
