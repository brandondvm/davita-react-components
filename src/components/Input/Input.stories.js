import React from 'react';
import Input from './Input';
import ReadOnlyInput from '../ReadOnlyInput/ReadOnlyInput';

export default {
  title: 'Input',
  component: Input,
};

let Template = (args) => <div className="storybook-wrapper"><Input {...args} /></div>;

export const Default = Template.bind({});
Default.args = {
  id: 'default',
  placeholder: 'Default'
};

export const InnerButtons = Template.bind({});
InnerButtons.args = {
  id: 'select-buttons',
  placeholder: 'Inner Buttons',
  label: 'Label',
  selectButtons: {
    leftButtonLabel: 'lbs',
    leftButtonOnClick: () => {},
    rightButtonLabel: 'kg',
    rightButtonOnClick: () => {}
  }
};

export const InnerLabel = Template.bind({})
InnerLabel.args = {
  id: 'inner-label',
  innerLabel: 'Units',
  placeholder: 'Placeholder',
  label: 'Label',
};

export const Max = Template.bind({});
Max.args = {
  id: 'max',
  placeholder: 'Max Characters',
  characterMax: '20'
};

export const Error = Template.bind({})
Error.args = {
  id: 'input-error',
  placeholder: 'Error',
  errorMessage: 'Error message',
};

Template = (args) => <div className="storybook-wrapper"><ReadOnlyInput {...args} /></div>;
export const ReadOnly = Template.bind({})
ReadOnly.args = {
  value: 'Read only text',
  label: 'Read Only',
};
