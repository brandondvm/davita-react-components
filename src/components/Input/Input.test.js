import enzyme from 'enzyme';
import React from 'react';

import Input from './Input';

const mockProps = {
  label: '',
  placeholder: '',
  onChange: jest.fn(),
  innerLabel: '',
  selectButtons: undefined,
  errorMessage: undefined,
  characterMax: undefined,
  id: 'input-id',
};

function setup(passedProps = {}) {
  const updatedProps = {
    ...mockProps,
    ...passedProps,
  };

  return enzyme.shallow(<Input { ...updatedProps } />);
}

describe('Input Component', () => {
  test('Component renders correctly simple empty field', () => {
    const wrapper = setup();
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  test('Component renders correctly with label', () => {
    const wrapper = setup({ label: 'Label' });

    expect(wrapper.get(0)).toMatchSnapshot();
  });

  test('Component renders correctly with select buttons', () => {
    const wrapper = setup({ selectButtons: {
      leftButtonLabel: 'lbs',
      leftButtonOnClick: jest.fn,
      rightButtonLabel: 'kg',
      rightButtonOnClick: jest.fn,
    } });

    expect(wrapper.get(0)).toMatchSnapshot();
  });

  test('Component renders correctly with error', () => {
    const wrapper = setup({
      label: 'Label',
      errorMessage: 'Error message',
    });

    expect(wrapper.get(0)).toMatchSnapshot();
  });

  test('Component renders correctly with label', () => {
    const wrapper = setup({
      label: 'Label',
      characterMax: '20',
      value: 'Some value'
    });

    expect(wrapper.get(0)).toMatchSnapshot();
  });
});
