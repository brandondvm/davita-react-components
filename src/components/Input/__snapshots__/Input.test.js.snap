// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component Component renders correctly simple empty field 1`] = `
<div
  className="input-component"
>
  <div
    className="input-component_input-area"
  >
    
    <label
      className="placeholder"
      htmlFor="input-id"
      style={
        Object {
          "color": false,
        }
      }
    >
      
    </label>
    <input
      aria-invalid={false}
      className="input-component_input   "
      id="input-id"
      inputMode="text"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      type="text"
    />
    
  </div>
  <div
    className="input-component_feedback"
  />
</div>
`;

exports[`Input Component Component renders correctly with error 1`] = `
<div
  className="input-component"
>
  <div
    className="input-component_input-area"
  >
    <label
      htmlFor="input-id"
      style={
        Object {
          "color": "#CE3616",
        }
      }
    >
      Label
    </label>
    <input
      aria-invalid={true}
      className="input-component_input error-present  "
      id="input-id"
      inputMode="text"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      type="text"
    />
    
  </div>
  <div
    className="input-component_feedback"
  >
    <span>
      Error message
    </span>
  </div>
</div>
`;

exports[`Input Component Component renders correctly with label 1`] = `
<div
  className="input-component"
>
  <div
    className="input-component_input-area"
  >
    <label
      htmlFor="input-id"
      style={
        Object {
          "color": "#0d1221",
        }
      }
    >
      Label
    </label>
    <input
      aria-invalid={false}
      className="input-component_input   "
      id="input-id"
      inputMode="text"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      type="text"
    />
    
  </div>
  <div
    className="input-component_feedback"
  />
</div>
`;

exports[`Input Component Component renders correctly with label 2`] = `
<div
  className="input-component"
>
  <div
    className="input-component_input-area"
  >
    <label
      htmlFor="input-id"
      style={
        Object {
          "color": "#0d1221",
        }
      }
    >
      Label
    </label>
    <input
      aria-invalid={false}
      className="input-component_input   "
      id="input-id"
      inputMode="text"
      maxLength="20"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      type="text"
    />
    
  </div>
  <div
    className="input-component_feedback"
  >
    <span
      className="input-component_char-counter"
    >
      0
      /
      20
    </span>
  </div>
</div>
`;

exports[`Input Component Component renders correctly with select buttons 1`] = `
<div
  className="input-component"
>
  <div
    className="input-component_input-area"
  >
    
    <label
      className="placeholder"
      htmlFor="input-id"
      style={
        Object {
          "color": false,
        }
      }
    >
      
    </label>
    <input
      aria-invalid={false}
      className="input-component_input   select-buttons-present"
      id="input-id"
      inputMode="text"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      type="text"
    />
    
    <div
      className="input-component_input-area_buttons"
    >
      <button
        className="active-button"
        onClick={[Function]}
        type="button"
      >
        lbs
      </button>
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        kg
      </button>
    </div>
  </div>
  <div
    className="input-component_feedback"
  />
</div>
`;
